geopmdpy/OPTIMIZER_README.md

```diff
@@ -0,0 +1,265 @@
+
+# Create evaluator
+evaluator = ApplicationEvaluator(
+    launch_command=["./benchmark"],
+    metric_regex="Score: ([0-9.]+)",
+    maximize=True,
+    timeout=600,
+    print_stdout=True
+)
+
+# Run optimization
+optimizer = BayesianOptimizer(grid, evaluator)
+result = optimizer.optimize(n_calls=50)
+
+print(f"Best metric: {result['best_metric']}")
+print(f"Best config: {result['best_config']}")
+```
+
+## Metric Extraction
+
+The optimizer uses Python regular expressions to extract performance metrics from application output:
+
+```bash
+--metric-regex "Performance: ([0-9.]+)"
+```
+
+The regex should capture the numeric metric value in a group. Examples:
+- `"GFLOPS: ([0-9.]+)"` - Extracts floating point after "GFLOPS: "
+- `"Time: ([0-9]+) seconds"` - Extracts integer time value
+- `"Score: ([0-9]*\.?[0-9]+)"` - Extracts decimal numbers with optional decimal point
+
+## Configuration Application
+
+The optimizer applies configurations using the GEOPM PIO interface:
+
+1. **Push Controls**: Each parameter is pushed to the system
+2. **Adjust Values**: Parameter values are set based on optimization coordinates
+3. **Write Batch**: All changes are applied atomically
+4. **Launch Application**: The application runs with the new configuration
+5. **Extract Metric**: Performance metric is parsed from application output
+
+## Error Handling
+
+The optimizer handles various failure modes:
+
+- **Timeout**: Applications that exceed the specified timeout are terminated
+- **Configuration Errors**: Invalid control parameters are caught early
+- **Regex Failures**: Missing or invalid metric patterns are reported
+- **Application Failures**: Non-zero exit codes and stderr output are captured
+
+## Best Practices
+
+1. **Start Small**: Begin with `--n-calls 20` to test your setup
+2. **Validate Regex**: Test your metric regex on sample application output
+3. **Choose Appropriate Domains**: Match control domains to distinct application behavior
+4. **Monitor Progress**: Watch the console output to verify convergence
+5. **Use Seeds**: Set `--random-seed` for reproducible experiments
+6. **Baseline First**: Run without optimization to establish baseline metrics
+7. **Set Appropriate Timeouts**: Use `--application-timeout` for long-running applications
+8. **Debug Systematically**: Use `--print-stdout` and higher verbosity levels when troubleshooting
+
+## Troubleshooting
+
+### Common Issues
+
+1. **Import Error**: Install scikit-optimize: `pip install scikit-optimize`
+2. **No Metric Found**: Check that your regex matches the application output
+3. **Configuration Failures**: Verify GEOPM service is running and accessible
+4. **Slow Convergence**: Increase `--n-calls` or check parameter ranges
+5. **Timeout Issues**: Adjust `--application-timeout` for your application's runtime
+
+### Debug Mode
+
+Set the `GEOPM_DEBUG` environment variable to see detailed error information:
+
+```bash
+GEOPM_DEBUG=1 python -m geopmdpy.optimizer ...
```

Comment:
Consider using log level to determine if exceptions are printed rather than the GEOPM_DEBUG environment variable.